queryable-spec.ts

```diff
@@ -0,0 +1,421 @@
+ *   applied.
+ * 
+ * If parameters of type Variable with an equal variable name are in place,
+ * then the corresponding quad components in the resulting quad stream MUST be
+ * equal.
+ * Expression's MAY contain Variable Term's. If their variable names are equal
+ * to Variable's in the given quad pattern, then the Expression MUST be 
+ * instantiated for each variable's binding in the resulting quad stream when
+ * applying the Expression filter.
+ */
+interface FilterableSource {
+  matchExpression(
+    subject?: RDF.Term,
+    predicate?: RDF.Term,
+    obj?: RDF.Term,
+    graph?: RDF.Term,
+    expression?: Expression,
+    opts?: { 
+      limit?: number; 
+      offset?: number; 
+    },
+  ): FilterableResult;
+};
+
+
+/******************************************************************************
+                              QUERYABLE SOURCE
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
+interface Bindings {
+  type: 'bindings';
+  has: (key: RDF.Variable) => boolean;
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  set: (key: RDF.Variable, value: RDF.Term) => Bindings;
+  delete: (key: RDF.Variable) => Bindings;
+  keys: () => Iterator<RDF.Variable>;
+  values: () => Iterator<RDF.Term>;
+  entries: () => Iterator<[RDF.Variable, RDF.Term]>;
+  forEach: (fn: (value: RDF.Term, key: RDF.Variable) => any) => void;
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings: (entries?: [RDF.Variable, RDF.Term][]) => Bindings;
+  filter: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => boolean) => Bindings;
+  map: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => RDF.Term) => Bindings;
+  merge: (left: Bindings, right: Bindings) => Bindings;
+  mergeWith: (
+    merger: (left: RDF.Term, right: RDF.Term, key: RDF.Variable) => RDF.Term,
+    left: Bindings,
+    right: Bindings,
+  ) => Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface QueryableResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+  cardinality?: number;
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+  [key: string]: any;
+}
+
+interface BaseQueryableResult {
```

Comment:
I see this pattern quite often in the wild, presumably because it makes adding properties that are common to all subtypes a bit easier. No strong opinions, happy to remove it and re-add later if needed.